editor/src/clj/editor/model.clj

```diff
@@ -137,43 +166,254 @@
+  (output gpu-textures g/Any :cached produce-gpu-textures)
+  (output dep-build-targets g/Any (gu/passthrough dep-build-targets))
+
+  (output scene-info g/Any (g/fnk [shader vertex-space gpu-textures name :as info] info))
+  (output material-binding-info g/Any (g/fnk [_node-id name material ^:try texture-binding-infos :as info]
+                                        (if (g/error-value? texture-binding-infos)
+                                          (assoc info :texture-binding-infos [])
+                                          info))))
+
+(defn- update-materials [evaluation-context self old-value new-value]
+  (when-let [{:keys [added removed changed renamed]} (util/order-agnostic-diff-by-map-key old-value new-value :name)]
+    (let [old-material-binding-infos (g/node-value self :material-binding-infos evaluation-context)
+          old-material-binding-name+order->info (util/multi-index-by-map-key old-material-binding-infos :name)]
+      (concat
+        (for [[[name] {:keys [material textures]}] added]
+          (g/make-nodes (g/node-id->graph-id self) [binding [MaterialBinding :name name]]
+            ;; We need to ensure material is applied first, so we don't lose the textures
+            ;; since model sampler names are matched against the material-defined samplers
+            (g/set-property binding :material material)
+            (g/set-property binding :textures textures)
+            (g/connect binding :_node-id self :nodes)
+            (g/connect binding :dep-build-targets self :dep-build-targets)
+            (g/connect binding :scene-info self :scene-infos)
+            (g/connect binding :material-binding-info self :material-binding-infos)))
+        (for [[old-material-binding-name+order [_ {:keys [material textures]}]] changed
+              :let [binding (:_node-id (old-material-binding-name+order->info old-material-binding-name+order))]
+              tx [(g/set-property binding :material material)
+                  (g/set-property binding :textures textures)]]
+          tx)
+        (for [[old-material-binding-name+order [new-material-binding-name]] renamed]
+          (g/set-property (:_node-id (old-material-binding-name+order->info old-material-binding-name+order))
+            :name new-material-binding-name))
+        (for [[material-binding-name+order] removed]
+          (g/delete-node (:_node-id (old-material-binding-name+order->info material-binding-name+order))))))))
+
+(defn- add-material [model-node-id material-name material-resource]
+  (g/update-property model-node-id :materials conj {:name material-name
+                                                    :material material-resource
+                                                    :textures []}))
+
+(defn- remove-material [model-node-id material-name]
+  (g/update-property model-node-id
+                     :materials
+                     (fn [materials]
+                       (filterv #(not= material-name (:name %)) materials))))
```

Comment:
My point was that we do not need to have this state in a property. We only used it for the diffing, and we don't need to diff, given that we either have no `old-value` (when loading), or know what happened (when adding or removing).

I think we should get rid of it, since it introduces the aforementioned concern of no single source of truth, potentially causing confusion and bugs in the future.

We could still have the `update-materials` function (perhaps renamed to `setup-materials`?) but call it at load time, without the diffing, and get rid of the `materials` property.